generator/mavgen.py

```diff
@@ -64,38 +64,82 @@ def mavgen(opts, args):
-
-                # Only parse new include files
-                if fname in all_files:
-                    continue
-                all_files.add(fname)
-
-                # Validate XML file with XSD file if possible.
-                if opts.validate:
-                    print("Validating %s" % fname)
-                    if not mavgen_validate(fname):
-                        return False
-                else:
-                    print("Validation skipped for %s." % fname)
-
-                # Parsing
-                print("Parsing %s" % fname)
-                xml.append(mavparse.MAVXML(fname, opts.wire_protocol))
-
-                # include message lengths and CRCs too
-                x.message_crcs.update(xml[-1].message_crcs)
-                x.message_lengths.update(xml[-1].message_lengths)
-                x.message_min_lengths.update(xml[-1].message_min_lengths)
-                x.message_flags.update(xml[-1].message_flags)
-                x.message_target_system_ofs.update(xml[-1].message_target_system_ofs)
-                x.message_target_component_ofs.update(xml[-1].message_target_component_ofs)
-                x.message_names.update(xml[-1].message_names)
-                x.largest_payload = max(x.largest_payload, xml[-1].largest_payload)
+        """Expand includes. XML files from arguments already parsed inot xml list."""
+        
+        def expand_oneiteration():
+            noincludeadded = True
+
+            for x in xml[:]:
+                try:
+                    if not x.parent:
+                        pass
+                except:
+                    x.parent=set()
+
+                for i in x.include:
+                    fname = os.path.join(os.path.dirname(x.filename), i)
+                    # Only parse new include files
+                    if fname in all_files:
+                        continue
+                    all_files.add(fname)
+                    # Validate XML file with XSD file if possible.
+                    if opts.validate:
+                        print("Validating %s" % fname)
+                        if not mavgen_validate(fname):
+                            print("ERROR Validation of %s failed" % fname)
+                            exit(1)
+                    else:
+                        print("Validation skipped for %s." % fname)
+
+                    # Parsing
+                    print("Parsing %s" % fname)
+
+                    parsed_xml = mavparse.MAVXML(fname, opts.wire_protocol)
+                    #add all parent(s) into included file
+                    parsed_xml.parent=set()
+                    parsed_xml.parent.add(x.basename)
+                    parsed_xml.parent.update(x.parent)
+                    xml.append(parsed_xml)
+
+                    noincludeadded = False
+            return noincludeadded        
+                    
+        for i in range(MAXIMUM_INCLUDE_FILE_NESTING):
```

Comment:
Throwing an error if we go too far would seem to be a good idea.

Reply:
We can't go too far - `MAXIMUM_INCLUDE_FILE_NESTING` is the maximum expansion.  

Not sure if you really mean throw, so have added this code on next line:
```
        if expand_oneiteration():
            print("ERROR Nesting level greater than maximum: %s" % MAXIMUM_INCLUDE_FILE_NESTING)
            exit(1)
```
So basically if there is more expansion that can be done this will do it and exit. Otherwise this will run but do nothing (no new files).